libfastboot/keybox_provision.c

```diff
@@ -16,202 +16,77 @@
-	UINTN remain;
-	UINT8 *data_addr = start_kb_addr;
-	UINT8 rpmb_key[RPMB_KEY_SIZE];
-
-	blk_cnt = write_sz / RPMB_BLOCK_SIZE;
-	remain = write_sz % RPMB_BLOCK_SIZE;
-	ret = get_rpmb_key(rpmb_key);
-	if (EFI_ERROR(ret)) {
-		efi_perror(ret, L"Failed to get keybox data");
-		goto exit;
-	}
-	ret = write_rpmb_data(NULL, blk_cnt, blk_addr, data_addr, rpmb_key, &rpmb_result);
-	debug(L"ret=%d, rpmb_result=%d", ret, rpmb_result);
+	ret = gpt_get_partition_by_label(L"teedata", &gpart, LOGICAL_UNIT_USER);
 	if (EFI_ERROR(ret)) {
-		efi_perror(ret, L"Failed to write keybox data");
+		efi_perror(ret, L"Partition teedata not found");
 		goto exit;
 	}
-	blk_addr += blk_cnt;
-	data_addr = data_addr + (blk_cnt * RPMB_BLOCK_SIZE);
-
-	if (remain) {
-		memset(rpmb_buffer, 0, RPMB_BLOCK_SIZE);
-		ret = memcpy_s(rpmb_buffer, sizeof(rpmb_buffer), data_addr, remain);
-		if (EFI_ERROR(ret))
-			goto exit;
-		ret = write_rpmb_data(NULL, 1, blk_addr, data_addr, rpmb_key, &rpmb_result);
-		debug(L"ret=%d, rpmb_result=%d", ret, rpmb_result);
-		if (EFI_ERROR(ret)) {
-			efi_perror(ret, L"Failed to write keybox data");
-			goto exit;
-		}
-	}
-exit:
-	//clear the rpmb key
-	memset(rpmb_key, 0, sizeof(rpmb_key));
-	barrier();
-#endif
-	return ret;
-}
 
-static EFI_STATUS write_attkb_data_sim(UINT8 *start_kb_addr, UINTN write_sz)
-{
-	EFI_STATUS ret = EFI_SUCCESS;
-	UINT16 blk_addr = ATTKB_META_BASE_ADDRESS;
-	UINT32 byte_offset = blk_addr * RPMB_BLOCK_SIZE;
+	partoffset = gpart.part.starting_lba * gpart.bio->Media->BlockSize;
 
-	ret = simulate_write_rpmb_data(byte_offset, start_kb_addr, write_sz);
-	debug(L"ret=%d", ret);
+	ret = uefi_call_wrapper(
+				gpart.dio->WriteDisk,
+				5,
+				gpart.dio,
+				gpart.bio->Media->MediaId,
+				partoffset + KB_HEAD_OFFSET,
+				sizeof(keybox_header_t),
+				(void *)&kb_header);
 	if (EFI_ERROR(ret)) {
-		efi_perror(ret, L"Failed to write keybox data");
-		return ret;
-	}
-
-	return ret;
-}
-
-EFI_STATUS flash_keybox(VOID *data, UINTN size)
-{
-	EFI_STATUS ret = EFI_SUCCESS;
-	UINTN total_sz;
-	UINTN write_sz;
-	UINTN kb_meta_sz;
-	UINT8 *cipher_blob;
-	UINTN cipher_blob_sz;
-	UINT8 *addr;
-
-	if (size > MAX_KEYBOX_SIZE) {
-		error(L"keybox size exceeded limit");
-		return EFI_INVALID_PARAMETER;
+		efi_perror(ret, L"Could not write to Disk.\n");
+		goto exit;
 	}
 
-	total_sz = sizeof(attkb_meta_block_t) + sizeof(encrypted_attkb_t) + size + GCM_TAG_SIZE;
-	start_kb_addr = AllocatePool(total_sz);
-        if (!start_kb_addr)
-                return EFI_OUT_OF_RESOURCES;
-
-	cipher_blob = start_kb_addr + sizeof(attkb_meta_block_t) + sizeof(encrypted_attkb_t);
-	ret = encrypt_keybox(data, size, cipher_blob, &cipher_blob_sz);
+	ret = uefi_call_wrapper(
+				gpart.dio->WriteDisk,
+				5,
+				gpart.dio,
+				gpart.bio->Media->MediaId,
+				sizeof(keybox_header_t) + partoffset + KB_HEAD_OFFSET,
+				size,
+				data);
 	if (EFI_ERROR(ret)) {
-		error(L"keybox encryption failure!");
+		efi_perror(ret, L"Could not write to Disk.\n");
```

Comment:
Suggest to use: "Could not write keybox data to Disk" to locate error message easily

Reply:
Never add `\n` with `debug()`, `error()` or `efi_perror()` as these macro already append the `\n`.